Enable library validation for webpushd

bnham opened the pull request:
#### de2a821c5ed5341d1f204053840f76571611e1f3
<pre>
Enable library validation for webpushd
<a href="https://bugs.webkit.org/show_bug.cgi?id=281869">https://bugs.webkit.org/show_bug.cgi?id=281869</a>
<a href="https://rdar.apple.com/137709472">rdar://137709472</a>

Reviewed by Elliott Williams.

The webpushd that is bundled in Safari updates for downlevel macOS uses the wrong version of
WebKit. Instead of using the WebKit bundled in the update, it uses the system WebKit. This breaks
Web Push since Safari tries to send messages to webpushd using a newer IPC version than the one that
webpushd&apos;s system WebKit expects.

The reason for this is that library validation isn&apos;t enabled on webpushd, which is a prerequisite
for webpushd picking the updated WebKit rather than the system WebKit.

* Source/WebKit/Configurations/webpushd.xcconfig:

Canonical link: <a href="https://commits.webkit.org/285571@main">https://commits.webkit.org/285571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07081d465578ac8b4e006904e06b1ffac2bf9214

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73080 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25888 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77295 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24318 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/291 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57437 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15920 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62894 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37854 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44089 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22647 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65947 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20719 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78959 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8 "") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65879 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62896 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65156 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16105 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7144 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/365 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3096 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/394 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/395 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

davidquesada commented on Source/WebKit/Configurations/webpushd.xcconfig:
```diff
@@ -41,8 +41,11 @@ CODE_SIGN_ENTITLEMENTS[sdk=*simulator] = Resources/ios/XPCService-embedded-simul
 
 WK_PROCESSED_XCENT_FILE=$(TEMP_FILE_DIR)/$(FULL_PRODUCT_NAME).entitlements
 
+WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS = $(WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_$(WK_LIBRARY_VALIDATION_ENABLED))
```
Where does `WK_LIBRARY_VALIDATION_ENABLED` ever get enabled? For XPC Services, it looks like `BaseXPCService.xcconfig` would define that as YES if it starts out as empty: https://github.com/WebKit/WebKit/blob/45e2f437b91c9aaeb1aa984bb6e4404e9f3fddac/Source/WebKit/Configurations/BaseXPCService.xcconfig#L96

but it's not clear to me where this would come from for this target, which isn't an XPC service.

bnham replied:
Yeah I was under the mistaken presumption that we set `WK_LIBRARY_VALIDATION_ENABLED` externally from as a project override in e.g. projects.json, but I guess you are right in that we just default library validation to enabled unless it's forced off via those lines in `BaseXPCService.xcconfig`. So I need to move that logic up to a more base xcconfig.

emw-apple commented on Source/WebKit/Configurations/BaseTarget.xcconfig:
```diff
@@ -76,3 +76,8 @@ WK_CORE_PREDICTION_DEFINES_YES = HAVE_CORE_PREDICTION;
+WK_LIBRARY_VALIDATION_ENABLED = $(WK_LIBRARY_VALIDATION_ENABLED_$(WK_EMPTY_$(WK_LIBRARY_VALIDATION_ENABLED)));
+WK_LIBRARY_VALIDATION_ENABLED_ = $(WK_LIBRARY_VALIDATION_ENABLED_NO);
+WK_LIBRARY_VALIDATION_ENABLED_NO = $(WK_LIBRARY_VALIDATION_ENABLED);
+WK_LIBRARY_VALIDATION_ENABLED_YES = YES;
```
```suggestion
```

emw-apple commented on Source/WebKit/Configurations/webpushd.xcconfig:
```diff
@@ -41,8 +41,11 @@ CODE_SIGN_ENTITLEMENTS[sdk=*simulator] = Resources/ios/XPCService-embedded-simul
 WK_PROCESSED_XCENT_FILE=$(TEMP_FILE_DIR)/$(FULL_PRODUCT_NAME).entitlements
 
+WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS = $(WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_$(WK_LIBRARY_VALIDATION_ENABLED))
+WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_YES = -o library
```
```suggestion
WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS = $(WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_$(WK_LIBRARY_VALIDATION_ENABLED));
WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_ = -o library;
WK_LIBRARY_VALIDATION_CODE_SIGN_FLAGS_YES = -o library;
```

emw-apple reviewed (commented):
I think you can simplify this. If I understand correctly, `WK_LIBRARY_VALIDATION_ENABLED` should pass a `-o library` code sign flag when enabled, and it should be enabled by default. Because build settings can be undefined, you can do that without the declaration on BaseTarget:

bnham committed:
Enable library validation for webpushd
https://bugs.webkit.org/show_bug.cgi?id=281869
rdar://137709472

Reviewed by Elliott Williams.

The webpushd that is bundled in Safari updates for downlevel macOS uses the wrong version of
WebKit. Instead of using the WebKit bundled in the update, it uses the system WebKit. This breaks
Web Push since Safari tries to send messages to webpushd using a newer IPC version than the one that
webpushd's system WebKit expects.

The reason for this is that library validation isn't enabled on webpushd, which is a prerequisite
for webpushd picking the updated WebKit rather than the system WebKit.

* Source/WebKit/Configurations/webpushd.xcconfig:

Canonical link: https://commits.webkit.org/285571@main
